src/librustdoc/html/highlight.rs

```diff
@@ -134,151 +135,321 @@ fn can_merge(class1: Option<Class>, class2: Option<Class>, text: &str) -> bool {
+        current_class.pending_exit = true;
+    }
+
+    fn last_class(&self) -> Option<Class> {
+        self.open_classes.last().map(|c| c.class)
+    }
+
+    fn last_class_is_open(&self) -> bool {
+        if let Some(last) = self.open_classes.last() {
+            last.open
+        } else {
+            // If there is no class, then it's already open.
+            true
+        }
+    }
+
+    fn close_last_if_needed<W: Write>(&mut self, out: &mut W) {
+        if let Some(last) = self.open_classes.last()
+            && last.pending_exit
+            && last.open
+        {
+            last.close_tag(out);
+            self.open_classes.pop();
+        }
+    }
+
+    fn push<W: Write>(
+        &mut self,
+        out: &mut W,
+        href_context: &Option<HrefContext<'_, '_>>,
+        class: Option<Class>,
+        text: Cow<'_, str>,
+        needs_escape: bool,
+    ) {
+        // If the new token cannot be merged with the currently open `Class`, we close the `Class`
+        // if possible.
+        if !can_merge(self.last_class(), class, &text) {
+            self.close_last_if_needed(out)
+        }
+
+        let current_class = self.last_class();
+
+        // If we have a `Class` that hasn't been "open" yet (ie, we received only an `EnterSpan`
+        // event), we need to open the `Class` before going any further so the new token will be
+        // written inside it.
+        if class.is_none() && !self.last_class_is_open() {
+            if let Some(current_class_info) = self.open_classes.last_mut() {
+                let class_s = current_class_info.class.as_html();
+                if !class_s.is_empty() {
+                    write!(out, "<span class=\"{class_s}\">").unwrap();
+                }
+                current_class_info.open = true;
+            }
+        }
+
+        let current_class_is_open = self.open_classes.last().is_some_and(|c| c.open);
+        let can_merge = can_merge(class, current_class, &text);
+        let should_open_tag = !current_class_is_open || !can_merge;
+
+        let text =
+            if needs_escape { Either::Left(&EscapeBodyText(&text)) } else { Either::Right(text) };
+
+        let closing_tag =
+            string_without_closing_tag(out, &text, class, href_context, should_open_tag);
+        if class.is_some() && should_open_tag && closing_tag.is_none() {
+            panic!(
+                "called `string_without_closing_tag` with a class but no closing tag was returned"
+            );
+        } else if let Some(closing_tag) = closing_tag
+            && !closing_tag.is_empty()
+        {
+            // If this is a link, we need to close it right away and not open a new `Class`,
+            // otherwise extra content would go into the `<a>` HTML tag.
+            if closing_tag == "</a>" {
+                out.write_str(closing_tag).unwrap();
+            // If the current `Class` is not compatible with this one, we create a new `Class`.
```

Comment:
This seems redundant with the logic in enter_elem?

I think if `can_merge` was modified to also take into account the spans, we wouldn't need this (and that change would also make merging more robust in general).